### PR TITLE
[FIXED JENKINS-44361] Follow HTTP redirects while initiating CLI connection

### PR DESCRIPTION
**Pull request by olivergondza**
# Description

When user initiated CLI connection talking to URL that redirects to Jenkins, it can either work, fail with reasonable exception or fail with rather cryptic exception.

https://issues.jenkins-ci.org/browse/JENKINS-44361

This make sure redirect is followed every time and if that does not resolve to Jenkins it is reported without stacktrace in a way user will understand.

### Changelog entries

Proposed changelog entries:

* Entry 1: JENKINS-44361, Follow HTTP redirects while initiating CLI connection.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Link to JIRA ticket in description, if appropriate
- [x] Appropriate autotests or explanation to why this change has no tests

### CODE REVIEW

File: test/src/test/java/hudson/cli/CLITest.java [hunk 1]
```diff
@@ -149,4 +174,95 @@ private void doInterrupt(File jar, File home, FreeStyleProject p, String... mode
...
     }
 
+    @Test
+    public void reportNotJenkins() throws Exception {
+        home = tempHome();
+        grabCliJar();
+
+        for (String transport: Arrays.asList("-remoting", "-http", "-ssh")) {
+
+            String url = "http://jenkins.io";
```
Comment (keerthikorivi) [inline test/src/test/java/hudson/cli/CLITest.java]: Can we consider declaring the string as final and outside the for loop

Comment (olivergondza) [inline test/src/test/java/hudson/cli/CLITest.java]: I guess we can...

File: cli/src/main/java/hudson/cli/CLI.java [hunk 2]
```diff
@@ -318,6 +316,27 @@ protected CliPort getCliTcpPort(URL url) throws IOException {
...
+        if (c.getHeaderField("X-Hudson")==null && c.getHeaderField("X-Jenkins")==null)
+            throw new NotTalkingToJenkinsException(c);
+    }
+    /*package*/ static final class NotTalkingToJenkinsException extends IOException {
+        public NotTalkingToJenkinsException(String s) {
+            super(s);
+        }
+
+        public NotTalkingToJenkinsException(URLConnection c) {
+            super("There's no Jenkins running at " + c.getURL().toString());
```
Comment (oleg-nenashev) [inline cli/src/main/java/hudson/cli/CLI.java]: or Jenkins is not accessible (proxy, firewall, etc.)

File: cli/src/main/java/hudson/cli/CLI.java [hunk 3]
```diff
@@ -405,6 +424,9 @@ public void upgrade() {
     public static void main(final String[] _args) throws Exception {
         try {
             System.exit(_main(_args));
+        } catch (NotTalkingToJenkinsException ex) {
+            System.err.println(ex.getMessage());
+            System.exit(-1);
```
Comment (oleg-nenashev) [inline cli/src/main/java/hudson/cli/CLI.java]: `-2` and document exit codes somewhere? May be useful to CLI users

**Review by oleg-nenashev — APPROVED** [review]
👍 , just minor improvement proposals

File: cli/src/main/java/hudson/cli/CLI.java (hunk 3 above)
Comment (olivergondza) [inline cli/src/main/java/hudson/cli/CLI.java]: Right, looking at the table in https://issues.jenkins-ci.org/browse/JENKINS-32273, it seems like `3`.

File: test/src/test/java/hudson/cli/CLITest.java [hunk 4]
```diff
@@ -149,4 +174,95 @@ private void doInterrupt(File jar, File home, FreeStyleProject p, String... mode
...
+            int ret = new Launcher.LocalLauncher(StreamTaskListener.fromStderr()).launch().cmds(
+                    "java", "-Duser.home=" + home, "-jar", jar.getAbsolutePath(), "-s", url, transport, "-user", "asdf", "who-am-i"
+            ).stdout(baos).stderr(baos).join();
+
+            assertThat(baos.toString(), containsString("There's no Jenkins running at"));
+            assertNotEquals(0, ret);
+        }
+    }
+
+    @Test
```
Comment (jglick) [inline test/src/test/java/hudson/cli/CLITest.java]: `@Issue`?

File: test/src/test/java/hudson/cli/CLITest.java (hunk 1 above)
Comment (jglick) [inline test/src/test/java/hudson/cli/CLITest.java]: Best to avoid trying to make Internet connections during tests—will make them flaky.

**Review by jglick — APPROVED** [review]
Did not follow all of the details. The general change sounds right.

**Comment by oleg-nenashev** [comment]
@olivergondza do you plan to address the exit code concern?

**Comment by olivergondza** [comment]
All review comments addressed.

**Review by oleg-nenashev — APPROVED** [review]
👍

**Comment by daniel-beck** [comment]
hudson.cli.CLIActionTest

> `expected:<3> but was:<255>`

> `expected:<3> but was:<-1>`

Seems related.

**Comment by olivergondza** [comment]
All 3 test failures ware related. All fixed locally...

**Review by oleg-nenashev — APPROVED** [review]
the recent changes look good to me

**Review by daniel-beck — APPROVED** [review]

**Pull request merged**